src/ucp/core/ucp_context.c

```diff
@@ -864,63 +858,142 @@ static int ucp_is_resource_in_device_list(const uct_tl_resource_desc_t *resource
+
+    if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW) {
+        /* Enable the transport if UCX_TLS=tl_name, or alias={tl_name} and
+         * UCX_TLS=alias. */
+        if ((ret & UCP_TL_IN_TRANSPORTS_LIST_REGULAR) != 0) {
+            return 1;
+        }
+
+        /* Enable the transport as an auxiliary if UCX_TLS=tl_name:aux, or
+         * alias={tl_name} and UCX_TLS=alias:aux, or alias={tl_name:aux} and
+         * UCX_TLS=alias. */
+        if (((ret & UCP_TL_IN_TRANSPORTS_LIST_AUX_IN_MAIN) != 0) ||
+            ((ret & UCP_TL_IN_TRANSPORTS_LIST_AUX_IN_ALIAS) != 0)) {
+            *rsc_flags |= UCP_TL_RSC_FLAG_AUX;
+            return 1;
+        }
+
+        return 0;
     }
 
-    /* check aliases */
-    for (alias = ucp_tl_aliases; alias->alias != NULL; ++alias) {
-        /* If an alias is in the list and the transport belongs this alias,
-         * enable/disable the transport (according to the list mode)
-         */
-        alias_arr_count = ucp_tl_alias_count(alias);
-        snprintf(info, sizeof(info), " for alias '%s'", alias->alias);
-        dummy_mask      = 0;
-        tmp_rsc_flags   = 0;
-        tmp_tl_cfg_mask = 0;
-        dummy_flags     = 0;
-        if (ucp_tls_array_is_present(alias->tls, alias_arr_count, tl_name, info,
-                                     &tmp_rsc_flags, &dummy_mask)) {
-            if (ucp_config_is_tl_name_present(allow_list, alias->alias, 1,
-                                              &dummy_flags, &tmp_tl_cfg_mask)) {
-                if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW) {
-                    *rsc_flags   |= tmp_rsc_flags;
-                    *tl_cfg_mask |= tmp_tl_cfg_mask;
-                    return 1;
-                } else if (!(tmp_rsc_flags & UCP_TL_RSC_FLAG_AUX)) {
-                    /* Disable the transport only if it is not used in the alias
-                     * as an auxiliary transport. */
-                    return 0;
-                }
-            }
+    if (!ucs_string_set_contains(aux_tls, tl_name)) {
+        /* Disable the transport which is not used as an auxiliary, if
+         * UCX_TLS=^tl_name, or alias={tl_name} and UCX_TLS=^alias. */
+        if ((ret & UCP_TL_IN_TRANSPORTS_LIST_REGULAR) != 0) {
+            return 0;
         }
+
+        return 1;
+    }
+
+    /* Only explicit indication in the deny list can completely disable
+     * transport which can be used as an auxiliary.
+     * E.g: UCX_TLS=^tl_name,tl_name:aux, or alias={tl_name} and
+     * UCX_TLS=^alias,alias:aux. */
+    if (((ret & UCP_TL_IN_TRANSPORTS_LIST_REGULAR) != 0) &&
+        ((ret & UCP_TL_IN_TRANSPORTS_LIST_AUX_IN_MAIN) != 0)) {
+        return 0;
+    }
+
+    /* If alias1={tl_name1, tl_name2:aux} and alias2={tl_name2}, then
+     * UCX_TLS=^tl_name2 or UCX_TLS=^alias2 should not break the usage of
+     * tl_name1. */
+    if ((ret & UCP_TL_IN_TRANSPORTS_LIST_REGULAR) != 0) {
+        *rsc_flags |= UCP_TL_RSC_FLAG_AUX;
+        return 1;
```

Comment:
shouldn't it be return 0?
since we return 1 in line 989 anyway

Reply:
We should return 1 here, to enable transport as an auxiliary.